protected/humhub/services/MigrationService.php

```diff
@@ -0,0 +1,328 @@
+<?php
+
+/*
+ * @link      https://www.humhub.org/
+ * @copyright Copyright (c) 2023 HumHub GmbH & Co. KG
+ * @license   https://www.humhub.com/licences
+ */
+
+namespace humhub\services;
+
+use humhub\commands\MigrateController;
+use humhub\components\Application;
+use humhub\components\Event;
+use humhub\components\Module;
+use humhub\events\MigrationEvent;
+use humhub\interfaces\ApplicationInterface;
+use Throwable;
+use Yii;
+use yii\base\ActionEvent;
+use yii\base\Component;
+use yii\base\Controller;
+use yii\base\InvalidConfigException;
+use yii\base\Module as BaseModule;
+use yii\console\ExitCode;
+
+/**
+ * @since 1.16
+ */
+class MigrationService extends Component
+{
+    public const EVENT_AFTER_MIGRATION = 'afterMigration';
+    public const EVENT_BEFORE_MIGRATION = 'beforeMigration';
```

Comment:
I know I suggested having two events here earlier. But isn't it enough to only implement `AFTER_MIGRATE`?

Reply:
As you wish.

Currently the developer has the option to do both in `Module`:

```php

class Module {

public function migrate() {

  // code that is run before the migration, replaced with new EVENT_BEFORE_MIGRATION
  // would currently allow to change the migration path
 
  parent::migrate();

  // code that is run after the migration, replaced with new EVENT_AFTER_MIGRATION
}

}
```

But I have no use case for the EVENT_BEFORE_MIGRATION. happy to remove.

Reply:
EVENT_BEFORE_MIGRATION has been removed.